go/vt/vttablet/tabletmanager/vreplication/vcopier.go

```diff
@@ -364,3 +437,268 @@ func (vc *vcopier) fastForward(ctx context.Context, copyState map[string]*sqltyp
+				}
+				return newVCopierBasicCopier(true, vc.vr.dbClient.stats, tableName, dbClient), nil
+			},
+			getCopyInsertConcurrency(),
+			vc.vr.stats,
+			tableName,
+		)
+	}
+	return newVCopierBasicCopier(false, vc.vr.dbClient.stats, tableName, vc.vr.dbClient)
+}
+
+func (vbc *vcopierBasicCopier) BeforeUpdateCopyState() *event.Hooks {
+	return vbc.hooks["updateCopyState:before"]
+}
+
+func (vbc *vcopierBasicCopier) Copy(ctx context.Context, rows []*querypb.Row, lastpk *querypb.Row) error {
+	start := time.Now()
+
+	var err error
+
+	// Begin transaction and insert rows.
+	if err = vbc.vdbClient.Begin(); err != nil {
+		err = fmt.Errorf("error beginning copy-update-state transaction: %s", err.Error())
+		goto DONE
+	}
+
+	if _, err = vbc.insertRows(ctx, rows); err != nil {
+		err = fmt.Errorf("error inserting rows: %s", err.Error())
+		goto DONE
+	}
+
+	// Update copy state.
+	if err = vbc.updateCopyState(ctx, lastpk); err != nil {
+		err = fmt.Errorf("error updating _vt.copy_state: %s", err.Error())
+		goto DONE
+	}
+
+	// Commit.
+	if err = vbc.vdbClient.Commit(); err != nil {
+		err = fmt.Errorf("error commiting copy-update-state transaction: %s", err.Error())
+		goto DONE
+	}
+
+	vbc.lastCommittedPk = lastpk
+
+DONE:
+	// Update stats.
+	vbc.stats.CopyRowCount.Add(int64(len(rows)))
+	vbc.stats.QueryCount.Add("copy", 1)
+	vbc.stats.TableCopyRowCounts.Add(vbc.tableName, int64(len(rows)))
+	vbc.stats.TableCopyTimings.Add(vbc.tableName, time.Since(start))
+
+	// Handle errors.
+	if err != nil {
+		vbc.stats.ErrorCounts.Add([]string{"Copy"}, 1)
+	}
+
+	return err
+}
+
+func (vbc *vcopierBasicCopier) Error() error {
+	return nil
+}
+
+func (vbc *vcopierBasicCopier) IsOpen() bool {
+	return vbc.isOpen
+}
+
+func (vbc *vcopierBasicCopier) LastCommittedPk() *querypb.Row {
+	return vbc.lastCommittedPk
+}
+
+func (vbc *vcopierBasicCopier) Open(
+	copyStateUpdate *sqlparser.ParsedQuery,
+	pkfields []*querypb.Field,
+	tablePlan *TablePlan,
+) {
+	if vbc.isOpen {
+		return
+	}
+
+	vbc.copyStateUpdate = copyStateUpdate
+	vbc.isOpen = true
+	vbc.pkfields = pkfields
+	vbc.tablePlan = tablePlan
+}
+
+func (vbc *vcopierBasicCopier) Close() {
+	if !vbc.IsOpen() {
+		return
+	}
+
+	if vbc.closeDbClient {
+		vbc.vdbClient.Close()
+	}
+
+	vbc.isOpen = false
+}
+
+func (vbc *vcopierBasicCopier) insertRows(ctx context.Context, rows []*querypb.Row) (*sqltypes.Result, error) {
+	if !vbc.IsOpen() {
+		return nil, fmt.Errorf("failed to insert rows with vcopier basic copier: not open")
+	}
+
+	return vbc.tablePlan.applyBulkInsert(&vbc.sqlbuffer, rows, func(sql string) (*sqltypes.Result, error) {
+		return vbc.vdbClient.ExecuteWithRetry(ctx, sql)
+	})
+}
+
+func (vbc *vcopierBasicCopier) updateCopyState(ctx context.Context, lastpk *querypb.Row) error {
+	if !vbc.IsOpen() {
+		return fmt.Errorf("failed to update copy state with vcopioer basic copier: not open")
+	}
+
+	vbc.BeforeUpdateCopyState().Fire()
```

Comment:
We should wrap this in a select so as to handle context cancellations. Otherwise we will end up with orphan goroutines if the previous batches don't complete due to invalid contexts.